File: monitoring-py/monitoring.py
#!/usr/bin/python3
from telnetlib import Telnet
from re import findall
from time import sleep, time
from os import environ
from influxdb import InfluxDBClient

ROUTER = environ["ROUTER_IP_ADDRESS"]
USER = environ["ROUTER_LOGIN_USER"]
PASSWORD = environ["ROUTER_LOGIN_PASSWORD"]
PROMPT = environ["ROUTER_PROMPT"]

DB_NAME = environ["INFLUX_DB_NAME"]
DB_CONTAINER = environ["INFLUX_DB_ADDRESS"]
DB_PORT = environ["INFLUX_DB_PORT"]
DB_USER = environ["INFLUX_DB_USER"]
DB_PASSWORD = environ["INFLUX_DB_PASSWORD"]

MONITORING_INTERVAL = int(environ["MONITORING_INTERVAL"])
BANDWIDTH_SAMPLING_INTERVAL = 1 # 1sec -> bps int(environ["BANDWIDTH_SAMPLING_INTERVAL"])

"""
firmware version
RTX1200 Rev.10.01.65 (Tue Oct 13 12:23:48 2015)
"""

class RTXTelnet:
    """
    RTXTelnet: Telnet wrapper for RTX series
    """

    log = ""

    def __init__(self, router, username, password, port=23, prompt="", timeout=5, wait=0.5):
        self.router = router
        self.username = username
        self.password = password
        self.port = port
        self.prompt = "\r\n" + prompt + "> "
        self.timeout = timeout
        self.wait = wait
        self.connect()

    def connect(self):
        """
        connect(self): Connect to RTX via telnet
        """
        # セッション開通
        self.telnet = Telnet(self.router, port=self.port, timeout=self.timeout)
        self.telnet.read_very_eager() # プロンプトを出す
        self.telnet.write(b"\n") # ←ナマステ!(ﾊﾞｼｯ)今日は無名ログインスンナスンナスンナスンナスンナスンナスンナスンナ
        self.telnet.read_very_eager() # プロンプト飛ばす

        # ユーザー名とパスワードでログインする
        while True:
            pro = self.telnet.read_very_eager().decode()
            self.log += pro
            if pro.endswith(self.prompt):
                break
            elif pro.endswith("\r\nUsername: "):
                self.telnet.write(self.username.encode("ascii") + b"\n")
            elif pro.endswith("\r\nPassword: "):
                self.telnet.write(self.password.encode("ascii") + b"\n")
            sleep(self.wait)

        # ASCIIモードを適用
        self.telnet.write(b"console character ascii\n\n")
        sleep(self.wait * 2)
        while True:
            pro = self.telnet.read_very_eager().decode()
            self.log += pro
            if pro.endswith(self.prompt):
                break

    def disconnect(self):
        """
        disconnect(self): Disconnect from RTX
        """
        self.telnet.write(b"exit\n")
        self.log += self.telnet.read_very_eager().decode()
        self.telnet.close()

    def execute(self, cmd):
        """
        execute(self, cmd): Execute command in RTX
        """
        # プロンプトを用意
        while True:
            pro = self.telnet.read_very_eager().decode()
            self.log += pro
            if pro.endswith(self.prompt):
                break
            if not pro:
                self.telnet.write(b"\n")
            sleep(self.wait)

        # 実行
        self.telnet.write(cmd.encode("ascii") + b"\n")
        sleep(self.wait * 2)
        res = ""
        while True:
            res += self.telnet.read_very_eager().decode()
            if res.endswith(self.prompt):
                self.log += res
                res = res.replace(cmd + " \x08 \x08\r\n", "").replace("---more---\r            \r", "").replace(self.prompt, "")
                break
            elif res.endswith("---more---"):
                self.telnet.write(b" ")
            sleep(self.wait)
        return res

def post_influxdb(dbconn, measurement, field, value):
    request = [
        {
            "measurement": measurement,
            "fields": {
                field: value,
            }
        }
    ]
    print(request)
    dbconn.write_points(request)
    return True

def grep(pattern, text):
    return findall(pattern, text)

def lan_interfaces():
    # RTX1200はLAN1 LAN2 LAN3だった
    return ["1", "2", "3"]

def pp_interfaces(config):
    t = []

    for w in grep(r"pp select (\d+)", config):
        t.append(w)

    return sorted(set(t), key=t.index)

def dhcp_scopes(config):
    t = []

    for w in grep(r"dhcp scope (\d+)", config):
        t.append(w)

    return sorted(set(t), key=t.index)

def lan_interface_speed(config, num):
    val = grep(r"speed lan"+num+r" (\d+\w?)", config)

    if (not val) or (val == 0):
        val = unitstr2num("1000m")

    return val

def unitstr2num(text):
    val = int(grep(r"(\d+)", text)[0])
    unit = text[-1:].lower()

    if unit == "k":
        return int(val * 1000)
    elif unit == "m":
        return int(val * 1000 * 1000)
    else:
        return int(val)

def environment_mon():
    status = TN.execute("show environment")

    # uptime
    uptime = grep(r"Elapsed time from boot: (\d+)days (\d+):(\d+):(\d+)", status)[0]
    days = int(uptime[0])
    hours = int(uptime[1])
    minutes = int(uptime[2])
    seconds = int(uptime[3])
    uptime_sec = days * 24 * 60 * 60 + hours * 60 * 60 + minutes * 60 + seconds
    post_influxdb(DB, "uptime", "sec", uptime_sec)

    # cpu
    post_influxdb(DB, "cpu", "5sec", int(grep(r"(\d+)%\(5sec\)", status)[0]))
    post_influxdb(DB, "cpu", "1min", int(grep(r"(\d+)%\(1min\)", status)[0]))
    post_influxdb(DB, "cpu", "5min", int(grep(r"(\d+)%\(5min\)", status)[0]))

    # memory
    post_influxdb(DB, "memory", "now", int(grep(r"Memory: (\d+)%", status)[0]))

    # packet buffer
    post_influxdb(DB, "packet_buffer", "small", int(grep(r"(\d+)%\(small\)", status)[0]))
    post_influxdb(DB, "packet_buffer", "middle", int(grep(r"(\d+)%\(middle\)", status)[0]))
    post_influxdb(DB, "packet_buffer", "large", int(grep(r"(\d+)%\(large\)", status)[0]))
    post_influxdb(DB, "packet_buffer", "huge", int(grep(r"(\d+)%\(huge\)", status)[0]))

    # temperature
    post_influxdb(DB, "temperature", "now", int(grep(r"Inside Temperature\(C.\): (\d+)", status)[0]))

def nat_mon():
    status = TN.execute("show nat descriptor address")
    if grep(r"(\d+) used.", status):
        value = int(grep(r"(\d+) used.", status)[0])
    else:
        value = -1
    post_influxdb(DB, "nat", "entry", value)

def dhcp_mon(config):
    for i in dhcp_scopes(config):
        status = TN.execute("show status dhcp "+i)
        post_influxdb(DB, "dhcp"+i, "leased", int(grep(r"Leased: (\d+)", status)[0]))
        post_influxdb(DB, "dhcp"+i, "usable", int(grep(r"Usable: (\d+)", status)[0]))

def pp_traffic_mon(config, sec):
    for i in pp_interfaces(config):
        start_time = time()
        status1 = TN.execute("show status pp "+i)
        sleep(sec)
        status2 = TN.execute("show status pp "+i)
        running_time = time() - start_time
        if "Connected" in status1:
            rcv1 = int(grep(r"\[(\d+) octets?\]", status1)[0])
            snd1 = int(grep(r"\[(\d+) octets?\]", status1)[1])
            rcv2 = int(grep(r"\[(\d+) octets?\]", status2)[0])
            snd2 = int(grep(r"\[(\d+) octets?\]", status2)[1])

            post_influxdb(DB, "pp"+i, "receive", (rcv2 - rcv1) / running_time)
            post_influxdb(DB, "pp"+i, "transmit", (snd2 - snd1) / running_time)

            rcv_load = int(grep(r"Load:\s+(\d+).(\d+)%", status1)[0][0]) + int(grep(r"Load:\s+(\d+).(\d+)%", status1)[0][1]) / 10
            snd_load = int(grep(r"Load:\s+(\d+).(\d+)%", status1)[1][0]) + int(grep(r"Load:\s+(\d+).(\d+)%", status1)[1][1]) / 10

            post_influxdb(DB, "pp"+i, "receive_load", rcv_load)
            post_influxdb(DB, "pp"+i, "transmit_load", snd_load)

def lan_traffic_mon(config, sec):
    for i in lan_interfaces():
        start_time = time()
        status1 = TN.execute("show status lan"+i)
        sleep(sec)
        status2 = TN.execute("show status lan"+i)
        running_time = time() - start_time

        bandwidth = lan_interface_speed(config, i)

        snd1 = int(grep(r"\((\d+) octets?\)", status1)[0])
        rcv1 = int(grep(r"\((\d+) octets?\)", status1)[1])
        snd2 = int(grep(r"\((\d+) octets?\)", status2)[0])
        rcv2 = int(grep(r"\((\d+) octets?\)", status2)[1])

        post_influxdb(DB, "lan"+i, "receive", (rcv2 - rcv1) / running_time)
        post_influxdb(DB, "lan"+i, "transmit", (snd2 - snd1) / running_time)

        post_influxdb(DB, "lan"+i, "receive_load", ((rcv2 - rcv1) * 8 / running_time) / bandwidth)
        post_influxdb(DB, "lan"+i, "transmit_load", ((snd2 - snd1) * 8 / running_time) / bandwidth)

def main():
    """
    Main
    """
    while True:
        try:
            config = TN.execute("show config")
            environment_mon()
            nat_mon()
            dhcp_mon(config)
            pp_traffic_mon(config, BANDWIDTH_SAMPLING_INTERVAL)
            lan_traffic_mon(config, BANDWIDTH_SAMPLING_INTERVAL)
        except:
            print("failed to post")
        sleep(MONITORING_INTERVAL)

if __name__ == '__main__':
    TN = RTXTelnet(ROUTER, USER, PASSWORD, prompt=PROMPT, timeout=3, wait=0.2)
    DB = InfluxDBClient(DB_CONTAINER, DB_PORT, DB_USER, DB_PASSWORD, DB_NAME)
    main()

File: monitoring.lua
--[[
    monitor.lua - InfluxDB driver for YAMAHA RTX series
]]

-- Interface monitoring interval (1-864000 sec)
avr_sec = 1

-- InfluxDB Endpoint
influxdb = "http://192.168.100.2:8086/write?db=rtx1200-lua"

-- Post monitoring value to InfluxDB
function post_influxdb(measurement, field, value)
    local request_tbl = {
        url = influxdb,
        method = "POST",
        content_type = "application/json",
        post_text = string.format("%s %s=%s",measurement,field,value)
    }
    response_tbl = rt.httprequest(request_tbl)
    --print(request_tbl.post_text)
    --print(response_tbl.body)
end

-- Get environment statuses
function get_env_status(tbl)
    local rtn, str
    rtn, str = rt.command("show environment")

    if (rtn) and (str) then
        for k, v in pairs(tbl) do
            v.val = str:match(v.ptn)
            if (v.val) then
                v.val = tostring(v.val)
            end
        end
    else
        return nil, nil
    end

    return rtn, str
end

-- Get DHCP lease status
function get_dhcp_status()
    local rtn, str, i
    local l = {}
    local u = {}

    rtn, str = rt.command("show status dhcp")
    if (rtn) and (str) then
        i = 1
        for w in string.gmatch(str, "Leased%:%s+(%d+)") do
            l[i] = w
            i = i + 1
        end
        i = 1
        for w in string.gmatch(str, "Usable%:%s+(%d+)") do
            u[i] = w
            i = i + 1
        end
    end

    return l, u
end

-- Get NAT status
function get_nat_status()
    local rtn, str, num, ip

    rtn, str = rt.command("show nat descriptor address ")
    if (rtn) and (str) then
        --ip = str:match("ipcp/(%a+)")
        num = str:match("(%d+)%s+used.")
    end

    --if (not ip) then
    --  ip = "0.0.0.0"
    --end
    if (not num) then
        num = "-1"
    end

    --return ip, num
    return num
end

-- Get LAN interface speed
function lan_interface_speed(num)
    local rtn, str, val, rt_name
    local cmd = "show config"
    local ptn = "speed lan" .. tostring(num) .. " (%d+%a)"

    rtn, str = rt.command(cmd)
    if (rtn) and (str) then
        str = str:match(ptn)
        if (str) then
            val = unitstr2num(str)
        end
    end

    if (not val) or (val == 0) then
        val = unitstr2num("1000m")
    end

    return val
end

-- Convert bit string to decimal
function unitstr2num(str)
    local val, unit

    val = tonumber(str:match("%d+"))
    unit = str:sub(-1):lower()

    if (unit == "k") then
        val = val * 1000
    elseif (unit == "m") then
        val = val * 1000 * 1000
    else
        val = 0
    end

    return val
end

-- Get LAN interfaces
function lan_interfaces()
    local rtn, str, n
    local cmd = "show config"
    local ptn = "ip lan(.+) address"
    local t = {}

    rtn, str = rt.command(cmd)

    if (rtn) and (str) then
        n = 1
        for w in string.gmatch(str, ptn) do
            t[n] = w
            n = n + 1
        end
    end

    return t
end

-- Get PP interfaces
function pp_interfaces()
    local rtn, str, n
    local cmd = "show config"
    local ptn = "pp select (%d+)"
    local t = {}

    rtn, str = rt.command(cmd)

    if (rtn) and (str) then
        n = 1
        for w in string.gmatch(str, ptn) do
            t[n] = w
            n = n + 1
        end
    end

    return t
end

-- Get LAN load
function lan_load_info(num, sec)
    local rtn, str1, str2, rcv, snd, rcv_load, snd_load
    local t = {}
    local cmd = "show status lan" .. tostring(num)
    local ptn = "%((%d+)%s+octets?%)"

    rtn, str1 = rt.command(cmd)
    if (rtn) and (str1) then
        rt.sleep(sec)

        rtn, str2 = rt.command(cmd)
        if (rtn) and (str2) then
            str1 = str1 .. str2

            n = 1
            for w in string.gmatch(str1, ptn) do
                t[n] = w
                n = n + 1
            end

            if (t[1]) and (t[3]) then
                snd = (tonumber(t[3]) - tonumber(t[1])) / sec
            end
            if (t[2]) and (t[4]) then
                rcv = (tonumber(t[4]) - tonumber(t[2])) / sec
            end
        end
    end

    return rcv, snd
end

-- Get PP load
function pp_load_info(num, sec)
    local rtn, str1, str2, rcv, rcv_load, snd, snd_load, n
    local t = {}
    local cmd = "show status pp " .. tostring(num)
    local ptn1 = "Load%:%s+(%d+)%.%d+%%"
    local ptn2 = "%[(%d+) octets?%]"

    rtn, str1 = rt.command(cmd)
    if (rtn) and (str1) then
        n = 1
        for w in string.gmatch(str1, ptn1) do
            t[n] = w
            n = n + 1
        end
        if (t[1]) then
            rcv_load = tonumber(t[1])
        end
        if (t[2]) then
            snd_load = tonumber(t[2])
        end
    end

    if (rtn) and (str1) then
        rt.sleep(sec)

        rtn, str2 = rt.command(cmd)
        if (rtn) and (str2) then
            str1 = str1 .. str2

            n = 1
            for w in string.gmatch(str1, ptn2) do
                t[n] = w
                n = n + 1
            end

            if (t[1]) and (t[3]) then
                rcv = ((tonumber(t[3]) - tonumber(t[1])) * 8) / sec
            else
                rtn = false
            end
            if (t[2]) and (t[4]) then
                snd = ((tonumber(t[4]) - tonumber(t[2])) * 8) / sec
            else
                rtn = false
            end
        end
    end

    return rcv, rcv_load, snd, snd_load
end

-- environment pattern
local env_tbl = {
    cpu_5sec = {
        ptn = "(%d+)%%%(5sec%)",
        val = 0
    },
    cpu_1min = {
        ptn = "(%d+)%%%(1min%)",
        val = 0
    },
    cpu_5min = {
        ptn = "(%d+)%%%(5min%)",
        val = 0
    },
    mem = {
        ptn = "(%d+)%% used",
        val = 0
    },
    buf_small = {
        ptn = "(%d+)%%%(small%)",
        val = 0
    },
    buf_middle = {
        ptn = "(%d+)%%%(middle%)",
        val = 0
    },
    buf_large = {
        ptn = "(%d+)%%%(large%)",
        val = 0
    },
    buf_huge = {
        ptn = "(%d+)%%%(huge%)",
        val = 0
    },
    temp = {
        ptn = "Inside Temperature%(C.%): (%d+)",
        val = 0
    },
    uptime = {
        ptn = "Elapsed time from boot: (%d+)days",
        val = 0
    }
}

-- Main loop
function main()
    local i, str, rcv, rcv_load, snd, snd_load
    local l = {}
    local u = {}

    -- Post environment data
    get_env_status(env_tbl)
    post_influxdb("cpu", "5sec", env_tbl.cpu_5sec.val)
    post_influxdb("cpu", "1min", env_tbl.cpu_1min.val)
    post_influxdb("cpu", "5min", env_tbl.cpu_5min.val)
    post_influxdb("memory", "now", env_tbl.mem.val)
    post_influxdb("packet_buffer", "small", env_tbl.buf_small.val)
    post_influxdb("packet_buffer", "middle", env_tbl.buf_middle.val)
    post_influxdb("packet_buffer", "large", env_tbl.buf_large.val)
    post_influxdb("packet_buffer", "huge", env_tbl.buf_huge.val)
    post_influxdb("temperature", "now", env_tbl.temp.val)
    post_influxdb("uptime", "day", env_tbl.uptime.val)

    -- Post DHCP leased data
    l, u = get_dhcp_status()
    for i,v in ipairs(l) do
        post_influxdb("dhcp"..tostring(i), "leased", v)
    end

    -- Post DHCP usable data
    for i,v in ipairs(u) do
        post_influxdb("dhcp"..tostring(i), "usable", v)
    end

    -- Post NAT status
    post_influxdb("nat", "entry", get_nat_status())

    -- Post PP load
    for i in ipairs(pp_interfaces()) do
        rcv, rcv_load, snd, snd_load = pp_load_info(i, avr_sec)
        post_influxdb("pp"..tostring(i) ,"receive",rcv_per)
        post_influxdb("pp"..tostring(i),"transmit",snd_per)
        post_influxdb("pp"..tostring(i),"receive_byte",rcv)
        post_influxdb("pp"..tostring(i),"transmit_byte",snd)
    end

    -- Post LAN load
    for i in ipairs(lan_interfaces()) do
        rcv, snd, imax = lan_load_info(i, avr_sec)
        post_influxdb("lan"..tostring(i), "receive_byte",rcv)
        post_influxdb("lan"..tostring(i), "transmit_byte",snd)
        post_influxdb("lan"..tostring(i), "interface_speed",lan_interface_speed(i))
    end
end

main()
os.exit(0)